registry/session_v2.go

```diff
@@ -95,11 +98,43 @@ func (r *Session) GetV2ImageManifest(ep *Endpoint, imageName, tagName string, au
 		return nil, "", utils.NewHTTPRequestError(fmt.Sprintf("Server error: %d trying to fetch for %s:%s", res.StatusCode, imageName, tagName), res)
 	}
 
-	buf, err := ioutil.ReadAll(res.Body)
+	manifestBytes, err := ioutil.ReadAll(res.Body)
 	if err != nil {
 		return nil, "", fmt.Errorf("Error while reading the http response: %s", err)
 	}
-	return buf, res.Header.Get(DockerDigestHeader), nil
+
+	sig, err := libtrust.ParsePrettySignature(manifestBytes, "signatures")
+	if err != nil {
+		return nil, "", fmt.Errorf("error parsing signed manifest: %s", err)
+	}
+
+	payload, err := sig.Payload()
+	if err != nil {
+		return nil, "", fmt.Errorf("error retrieving signed manifest payload: %s", err)
+	}
+
+	hdrDigest, err := digest.ParseDigest(res.Header.Get(DockerDigestHeader))
+	if err != nil {
+		return nil, "", fmt.Errorf("invalid manifest digest from registry: %s", err)
+	}
+
+	dgstVerifier, err := digest.NewDigestVerifier(hdrDigest)
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to verify manifest digest from registry: %s", err)
+	}
+
+	dgstVerifier.Write(payload)
+
+	if !dgstVerifier.Verified() {
+		computedDigest, _ := digest.FromBytes(payload)
```

Comment:
It should be unaltered from what was pushed, yes. What's pushed is a manifest payload with attached signatures and the digest is only of the payload.

> On Mar 18, 2015, at 15:11, Andy Goldstein <notifications@github.com> wrote:
> 
> In registry/session_v2.go:
> 
> > +	}
> > +
> > +	hdrDigest, err := digest.ParseDigest(res.Header.Get(DockerDigestHeader))
> > +	if err != nil {
> > +		return nil, "", fmt.Errorf("invalid manifest digest from registry: %s", err)
> > +	}
> > +
> > +	dgstVerifier, err := digest.NewDigestVerifier(hdrDigest)
> > +	if err != nil {
> > +		return nil, "", fmt.Errorf("unable to verify manifest digest from registry: %s", err)
> > +	}
> > +
> > +	dgstVerifier.Write(payload)
> > +
> > +	if !dgstVerifier.Verified() {
> > +		computedDigest, _ := digest.FromBytes(payload)
> Is the payload guaranteed to be the same as what the registry is storing? Any chance the fields could show up in a different order?
> …
> —
> Reply to this email directly or view it on GitHub.
>